### ISSUE TITLE
required fields with default value being generated to a nullable field

### ISSUE THREAD

**Issue by subzero911** [heftapp/graphql_codegen#338]
<img width="282" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/fb2959c5-462c-4c81-9552-3dae49f99a74">

Expected conversion:
`required Enum_ColorMode? colorMode,`

Actual conversion:
`Enum_ColorMode? colorMode,`

`!` in graphql means "required", not "non-nullable". It can be both required and nullable.

I expect that if I have a required field with default value, than I MUST put some value on Dart side, but I CAN put just "null", and a server will substitute it with default value in this case.

related to https://github.com/heftapp/graphql_codegen/issues/293

**Comment by budde377** [comment]
Hi @subzero911 

Please provide some more context to allow me to efficiently help you.

This could be an example GraphQL operation and schema along with generated code and what you expect to be generated.

**Comment by subzero911** [comment]
My scheme:
<img width="748" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/069284ec-b1fd-4ab9-8a1c-21b51e5c0cca">

generated code:
<img width="516" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/2df7362f-6b10-4515-aa8a-af04bec73ccc">

expected to be generated...
<img width="505" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/9c68bee2-aa07-4bd8-997d-00a305cc4147">

So the compiler will force me to fill the value:
<img width="326" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/34a8572e-d1b4-4af5-8554-99a1cee1fcb7">

<img width="324" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/f39ddece-f810-46ae-9d44-118f9efb6708">

**Comment by budde377** [comment]
I don't see how this is not solved by #293?

**Comment by budde377** [comment]
I.e I don't agree that the expected code is in fact expected. If you omit the value or supply `null` it should not be sent to the server.

**Comment by budde377** [comment]
Yes, but the implementation differs. I.e null values will be omitted. I believe you confirmed with your backend dev eight months ago that this works by letting the sever fill out the default value?

**Comment by subzero911** [comment]
You're probably right.
GraphQL specs assume that only optional values can have a default value:
<img width="814" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/65e406bd-cf77-4126-bed1-f2144c6ca92a">

So it's rather an incorrect scheme than a bad code generation.

**Comment by subzero911** [comment]
Hi again, I checked this twice
There's a nuance between arguments and types:
- in field arguments `!` means required:
![image](https://github.com/heftapp/graphql_codegen/assets/12999702/8b1ee39b-5485-4eb7-85ca-a54c92aa6cb7)

- in fields `!` means non-nullable:
https://graphql.org/graphql-js/basic-types/
<img width="806" alt="image" src="https://github.com/heftapp/graphql_codegen/assets/12999702/d5c42277-9e55-42e7-99a3-9bb646b325ff">

So it's legit to write `colorMode: ColorMode! = 'light'`
In this case, code generation tool should generate:
`Enum_ColorMode colorMode = 'light',`
instead of `Enum_ColorMode? colorMode`

> I believe you confirmed with your backend dev eight months ago that this works by letting the sever fill out the default value?

It's unsafe to rely on server implementation.
It could lead to a scheme validation failure, if I sent null, but the server was waiting for a value

**Comment by budde377** [comment]
Thank you for your persistence on this issue and for keeping up a good debate.

According to your first image, the behaviour for the default value when you provide it to a required field is not defined. As such, you can argue that the engineer writing the schema should not add default values to required fields. Of course, this is not a perfect world, and I recognise that considering that schema for invalid when generating the client is not an option. 

I am hesitant to generate the default value because it currently isn't a concern for the generator, and making it so carries a few problems:

* *Backwards compatibility*: Right now the schema can update the default values as it see fit in a backwards compatible way. The server can change the default value, and the client will not need to be rebuilt because they are not generating and sending any default value. This is a very powerful feature.
* *Inconsistent default value behaviour*: Say that we only generate the default value for the required fields. This might limit the impact of the point above, but it will require an inconsistent handling of default values. It is not clear to the implementor when the client will send a default value and when it will not.
* *Complex default value serialisation*: Generating the Enum value is easy, but consider complex input types or custom scalars. If we start a best-effort attempt at parsing the default value from the schema, converting it into a Dart type, and setting it, we are opening a can of worms I am not sure I want to handle. 

With these reasons in mind, I think there are two appropriate solutions:

* Do not make the field required, and do not send it to the server (current implementation).
* Do make the field required and force the client-developer to provide a value (old implementation).

As you are pointing out, different servers might have different interpretations of the spec, so I think it would be reasonable if the solution was configurable. We can add a flag to the config allowing the user to specify their desired behaviour.
